### PR TITLE
Fix false positives on closure rules on Swift 2.3

### LINKED ISSUE #1019: False positive on closure_parameter_position with Swift 2.3
Hey,

the following code triggers a parameter position violation for the completion closure :

```
UIView.animateWithDuration(0.3, animations: {
  self.disableInteractionRightView.alpha = 0
  }, completion: { _ in
    self.disableInteractionRightView.hidden = true
})
```

Do you have any idea why ?

### PR DESCRIPTION
**Pull request by marcelofabri**
This should fix #1019 

Maybe we should have a way of running the tests on all Swift versions 🤔

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -52,6 +52,11 @@
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#928](https://github.com/realm/SwiftLint/issues/928)
 
+* Fix false positives on `closure_parameter_position` and 
+  `unused_closure_parameter` rules on Swift 2.3.
```
Comment (jpsim) [inline CHANGELOG.md]: "with Swift 2.x" instead of "on Swift 2.3"

File: Source/SwiftLintFramework/Extensions/Dictionary+SwiftLint.swift [hunk 2]
```diff
@@ -27,7 +27,13 @@ extension Dictionary where Key: ExpressibleByStringLiteral {
             }
 
             if SwiftDeclarationKind(rawValue: kindString) == .varParameter {
-                return [subDict]
+                let nestedParameters = subDict.enclosedVarParameters
```
Comment (jpsim) [inline Source/SwiftLintFramework/Extensions/Dictionary+SwiftLint.swift]: Shouldn't this switch on the current Swift version instead?

File: Source/SwiftLintFramework/Models/SwiftVersion.swift [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+//
+
+import SourceKittenFramework
+
+enum SwiftVersion {
+    case two
+    case three
+
+    static let current: SwiftVersion = {
+        let file = File(contents: "#sourceLocation()")
```
Comment (jpsim) [inline Source/SwiftLintFramework/Models/SwiftVersion.swift]: We should propose that a request type be added to SourceKit to get the current Swift version. But this is good for now.

File: Source/SwiftLintFramework/Extensions/Dictionary+SwiftLint.swift [hunk 4]
```diff
@@ -27,7 +27,13 @@ extension Dictionary where Key: ExpressibleByStringLiteral {
...
+                // on Swift 2.3, a closure parameter is inside another .varParameter and not inside an .argument
```
Comment (jpsim) [inline Source/SwiftLintFramework/Extensions/Dictionary+SwiftLint.swift]: Generally "with Swift X.Y", not "on Swift X.Y".

**Comment by marcelofabri** [comment]
@jpsim We can also create another scheme in Xcode changing the value of `TOOLCHAINS` env var (it's already being set to `${TOOLCHAINS}`). In that scheme we can disable the Swift 3 tests.

Besides the integration tests (that I've disabled), the only two currently failing tests are:
* `attributes` rule: there's some history on the PR that added this rule, but basically some attributes were just added on Swift 3
* `number_separator` rule: it seems that with Swift 3, SourceKit reports the negative sign as part of the number but Swift 2.3 doesn't, so it's only a matter of where the violation marker should be.

File: Source/SwiftLintFramework/Models/SwiftVersion.swift [hunk 5]
```diff
@@ -23,4 +23,11 @@ enum SwiftVersion {
 
         fatalError("Unexpected Swift version")
     }()
+
+    var nameKey: String {
```
Comment (jpsim) [inline Source/SwiftLintFramework/Models/SwiftVersion.swift]: This is now used in just one place, keep it there? This shouldn't be a widely used thing.

Comment (marcelofabri) [inline Source/SwiftLintFramework/Models/SwiftVersion.swift]: 👍

Comment (marcelofabri) [inline Source/SwiftLintFramework/Models/SwiftVersion.swift]: Actually this is being used on `enclosedVarParameters` as well 🤔

Comment (marcelofabri) [inline Source/SwiftLintFramework/Models/SwiftVersion.swift]: But I could use "key.typename" directly there, as it's used inside the switch. What do you think would be better?

Comment (jpsim) [inline Source/SwiftLintFramework/Models/SwiftVersion.swift]: I think `nameKey` isn't representative of the conditional value it returns. Copying the code inline avoids coming up with an accurate name for it. So either come up with a good name, or skirt around the issue :-)

Comment (marcelofabri) [inline Source/SwiftLintFramework/Models/SwiftVersion.swift]: I've created a private `nameKey(for version: SwiftVersion)` function where it's really needed. I hope it's a good enough name as it's private 😅

Comment (jpsim) [inline Source/SwiftLintFramework/Models/SwiftVersion.swift]: Yup!

**Pull request merged**